### ISSUE TITLE
TS can't automatically infer type from outer function in certain cases

### ISSUE THREAD

**Issue by fabian-hiller** [microsoft/TypeScript#57978]
### 🔎 Search Terms

infer function argument bug

### 🕗 Version & Regression Information

- I think this is a bug
- I could not find a change between versions

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIILMA9iAPANTgBsBXCAPmQG8AoZZAfShJAApQAHEsALmUNIQAlHwFkA3DQC+NGjBYZsIZAFtQAGQggA5mAAWBYmWQQAHpBAATAM7JrYKKG3IAPsjggAngG0AuuVY6VVA+EBIVACNoGhE0RRxDQUpaeigIMBIoZRT6RmY2Tm4haiDc5GAYZHYQLjAAOiItXT1kXGCQYpyy+n0oLAB3ZBAIQYBRKD6oVgByNRBNHX1poUlu5Bk1tIys8pruVdypABogqUkZGgB6S-5oayVkAEYq-RQcFH7gfXdlaEmhWTyEDxZQcYAcCCPXBBAAq6EwOGeZgsNjiCLw9kcOnIJwCiHRjz4cJBj1i1VqfExTmKAF5KFSdCVUulMspyUVkHT3CS6kwWOywCtpLIEDh7MgAMoIPQQFRwZ405BgiGPVhzBbNViklbIa7IGESux6LAkIiWXYwaDIAAGDO01uQMD6KmQr2QJsgUEdCnRsj1+DuDwATFUvsh+lgoABrWz9GXKfpQOAcCHm0DuV0kDiNAFyH0PZUQIPQ+jEgkmcxaVHeeFKXB28i+XGsfFKQn62uIsmFXh2BzUzn0-uMnJbVmhvaCwfcgneAAMvl5+QFQouopA4qlMrlIcVhaDrG86qa+i1Ql8Or1AAloCgwxHo7Y4LYzBCMBBLEA

### 💻 Code

```ts
interface Action<Value> {
  _run(input: Value): Value;
}

function minLength<Value extends string | any[]>(
  min: number
): Action<Value> {
  return {
    _run(input) {
      if (input.length < min) {
        throw new Error('minLength');
      }
      return input;
    },
  };
}

// Version 1 (the one with an error)

function pipe1<
  TAction1 extends Action<string>,
>(action1: TAction1): (input: string) => string {
  return (input) => action1._run(input);
}

const Schema1 = pipe1(minLength(1)); // TS should infer `string` from the outer function

// Version 2 (it works when wrapped in a tuple)

function pipe2<
  TAction1 extends [Action<string>],
>(action1: TAction1): (input: string) => string {
  return (input) => action1[0]._run(input);
}

const Schema2 = pipe2([minLength(1)]); // Here it works as expected
```

### 🙁 Actual behavior

In certain cases, TS will not infer the type of the inner function based on the outer function automatically and will display an error instead. In other cases, such as when the type is wrapped in a tuple, it works. That's why I think the current behavior is inconsistent. Maybe there's a reason for this that I'm not aware of, otherwise it could be a bug.

### 🙂 Expected behavior

I would expect TS to infer the type and not show an error. TS should have all the information it needs to infer the type automatically.

### Additional information about the issue

I discovered this problem while rewriting the source code of Valibot: https://valibot.dev/

**Comment by RyanCavanaugh** [comment]
Why isn't `pipe1` declared this way?
```ts
function pipe1(action1: Action<string>): (input: string) => string {
  return (input) => action1._run(input);
}
```

**Comment by fabian-hiller** [comment]
Because in the actual source code, `pipe` is a dynamic function that assigns the data type of an action based on the previous action. Without generics, this would require an "infinite" number of overload signatures to support any possible case.

**Comment by fabian-hiller** [comment]
Because in the actual source code, `pipe` is a dynamic function that assigns the data type of an action based on the previous action. Without generics, this would require an "infinite" number of overload signatures to support any possible case.

**Comment by Andarist** [comment]
duplicate of https://github.com/microsoft/TypeScript/issues/52864
https://github.com/microsoft/TypeScript/pull/52866 tries to address this and awaits further review

**Comment by fabian-hiller** [comment]
@RyanCavanaugh I looked at issue #52864. The problem is similar, but the code is completely different. Therefore, I would not consider this PR a "duplicate". There is also no evidence that PR #52866 actually fixes this issue. Could you re-evaluate it? Unfortunately, this problem blocks the entire [rewrite of Valibot](https://github.com/fabian-hiller/valibot/discussions/463).

**Comment by Andarist** [comment]
> There is also no evidence that PR https://github.com/microsoft/TypeScript/pull/52866 actually fixes this issue.

I already added the test case from this issue to my PR - roughly at the same time when I commented here that it's a duplicate of that other issue. See the commit with the test: https://github.com/microsoft/TypeScript/pull/52866/commits/71421577603dc230ea63de4ce03781103872a47e

> The problem is similar, but the code is completely different.

The root cause really seems to be the same so I think it's fair to classify this as a duplicate of the other issue. The fact that it doesn't work when ur inner call is directly at the argument position of the outer call while it works when it's deeper in the argument is a sign for me that it's the same issue (the other one suffered from **exactly** the same thing).
